pkg/kubelet/cm/deviceplugin/endpoint_handler.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deviceplugin
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha"
+)
+
+// endpointHandler handles all operations related to endpoints
+// namely creating the endpoints, keeping a map of them and stoping them
+type endpointHandler interface {
+	NewEndpoint(resourceName, socketPath string) error
+
+	Devices() map[string][]pluginapi.Device
+	Endpoint(resourceName string) (endpoint, bool)
+
+	Stop() error
+
+	Store() endpointStore
+	SetStore(endpointStore)
+
+	Callback() managerCallback
+	SetCallback(f managerCallback)
+}
+
+type endpointStore interface {
+	Endpoint(resourceName string) (endpoint, bool)
+	SynchronizedEndpoint(resourceName string) (*synchronizedEndpoint, bool)
+
+	SwapEndpoint(e endpoint) (endpoint, bool)
+	DeleteEndpoint(resourceName string) error
+
+	Range(func(string, *synchronizedEndpoint))
+}
+
+type synchronizedEndpoint struct {
+	e    endpoint
+	done chan string
+}
+
+type endpointStoreImpl struct {
+	sync.Mutex
+	endpoints map[string]*synchronizedEndpoint
+}
+
+type endpointHandlerImpl struct {
+	store    endpointStore
+	callback managerCallback
+}
+
+func newEndpointHandlerImpl(f managerCallback) *endpointHandlerImpl {
+	return &endpointHandlerImpl{
+		store:    newEndpointStoreImpl(),
+		callback: f,
+	}
+}
+
+func (h *endpointHandlerImpl) NewEndpoint(resourceName, socketPath string) error {
+	e, err := newEndpoint(socketPath, resourceName)
+	if err != nil {
+		return err
+	}
+
+	var devStore deviceStore = newDeviceStoreImpl(h.callback)
+	if old, ok := h.store.Endpoint(resourceName); ok && old != nil {
+		devStore = old.Store()
+	}
+
+	e.SetStore(devStore)
+
+	old, ok := h.store.SwapEndpoint(e)
+	go h.trackEnpoint(e)
+
+	if ok && old != nil {
+		old.SetStore(newAlwaysEmptyDeviceStore())
```

Comment:
Endpoint now has a `DeviceStore` so that we can move the device map to the new Endpoint when re-registration happen.

The `DeviceStore` also has a Callback method that is Called when an update happens or when the endpoint is stoped.
In case of re-registration, we want to stop the old endpoint but not call the Callback, which is why we set the store to `newAlwaysEmptyDeviceStore`.
The endpoint will call the `DeviceStore` Callback but an `AlwaysEmptyDeviceStore` will never call the Manager callback.